### DISCUSSION TITLE
Potential issues when deploying with capistrano

### DISCUSSION THREAD

**Discussion by rachitpant** [ElMassimo/vite_ruby#198] [General]
### Description 📖

Thank you for your help yesterday and for this gem in general. I am trying to deploy a rails application that I migrated from webpacker to vite to our staging environment. I was able to deploy a working copy but there are some issues with the contents of public directory on the server that I have tried to describe below. You will notice my `publicOutputDir` is `""`. I am not sure if the following are actually issues so I have started this discussion and not opened a bug.

### Vite Ruby Info
```
bin/vite present?: true
vite_ruby: 3.0.9
vite_rails: 3.0.6
rails: 6.1.4.4
node: v14.16.1
npm: 8.3.0
yarn: 1.22.17
pnpm: 
ruby: ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin21]
```
## Problem 1
When watched files have not changed and you try to deploy with capistrano, vite build would not build new assets and deploy existing contents within the public directory. There is a false assumption that the contents of last build would reside locally in public directory. Thats true only if you build locally with a command like `RAILS_ENV=staging DEBUG=vite-plugin-ruby:* bin/vite build`.
However when deploying with a capistrano deploy script , the assets are directly generated on the server and 
In this scenario, the public directory in the release folder on the deployment server will not include the manifest files.

## Problem 2
 If there are existing files in the public directory ( eg. robots.txt ) , those files would not get deployed on the server in case of a successful vite build. Please see the screenshot below. You can see the manifest files and the assets generated by vite , but the existing public files are not present.
<img width="912" alt="image" src="https://user-images.githubusercontent.com/1474415/159527199-8d1f6099-514a-4df5-874d-d355faaeb04c.png">

On the other hand when vite doesn't actually build because watch files haven't changed, you can find the existing public files on the server but no manifest files. ( essentially this is same as problem 1 , described above)
<img width="912" alt="Screen Shot 2022-03-22 at 11 42 36 AM" src="https://user-images.githubusercontent.com/1474415/159526422-1cd9cf5e-0e21-4c97-a14c-149bd3a85686.png">

### Ideal scenario
Ideally, both the existing static files on the src public folder (eg. robots.txt ) as well as manifest files should get generated every time on the server public directory irrespective of whether the watched files have changed or not. It shouldn't be an either/or scenario. Right now the result is one of the above two screenshots.

## Problem 3.

When vite does a build ( either the first time or watch files have changed ) , it recreates the assets folder instead of following an existing symlink to the shared assets directory if there is one ( by default[ capistrano does this ](https://stackoverflow.com/questions/42673337/capistrano-what-are-the-default-linked-dirs-in-config-deploy-rb-for)) .
If you compare the two screenshots above, you will notice in the first screenshot vite-ruby replaced has replaced the assets symlink with a new assets folder and generated the assets there.

**Comment by ElMassimo** [comment]
About builds being skipped, you should ensure that either:
- The content of the `publicOutputDir` dir is persisted across builds (which is the case in most setups)
- The [`buildCacheDir`](https://vite-ruby.netlify.app/config/#buildcachedir) is cleared across builds if you don't intend to reuse previous assets

Regarding `robots.txt`, you can avoid that by moving `robots.txt` to `app/frontend/public`, or avoid using `"publicOutputDir": ""` (which is not recommended, it's better to keep the `"publicOutputDir": "vite"`  default).

About the third problem, if you avoid setting `"publicOutputDir": ""`, it wouldn't affect any symlinks already in place in `public`. 

---

I'm not available for further advice regarding the particulars of a Capistrano-based deployment.

**Reply by rachitpant** [reply]
> * The content of the `publicOutputDir` dir is persisted across builds (which is the case in most setups)

is there something that should be done so that build is outputted to `publicOutputDir` and its persisted. The two screenshots above are consecutive builds and the manifest file generated in the first build was not present in the second build.
